### PR TITLE
Initial commit for data collector code

### PR DESCRIPTION
**Pull request by rajusem**
Changes done so far as below. 
- Separated out big query collection code form inference pipeline
- Refactored few classes
- Reading AWS values/github credentials from cloud constants
- Containerised the codebase. 
- added templates for onetime job as well as cron job.

### CODE REVIEW

File: src/utils/bq_utils.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
+"""
```
Comment (arajkumar) [inline src/utils/bq_utils.py]: Can you refer to this package as a dependency instead of copying the implementation into our codebase?

https://github.com/SohierDane/BigQuery_Helper

File: scripts/entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+set +ex
+
+ecosystem_list=("openshift" "knative" "kubevirt")
+
+# Run the data collector for given eco systems and no of days to colelct data
+python3 app/run_data_collector.py -eco-systems ${ecosystem_list[*]} -days ${DAYS};
```
Comment (arajkumar) [inline scripts/entrypoint.sh]: Usual convention is to use double hyphen for expanded flag name, and single hyphen for single char flag name.

e.g. --ecosystem or -e

File: src/utils/aws_utils.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    resolved_path = Path(folder_path).resolve()
+    parent_dir = resolved_path.parent
+    for root, _, filenames in os.walk(resolved_path):
+        for filename in filenames:
+            if root != '.':
+                s3_dest = os.path.join(prefix,
+                                       Path(root).relative_to(parent_dir), filename)
+            else:
+                s3_dest = os.path.join(prefix, filename)
+            _logger.info('Uploading to: {d}'.format(d=s3_dest))
```
Comment (arajkumar) [inline src/utils/aws_utils.py]: This should be a debug print.

File: run_data_collector.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    _logger.info('Merging issues and pull requests datasets')
+    cols = issues_df.columns
+    data_frame = pd.concat([issues_df, prs_df], axis=0, sort=False, ignore_index=True).reset_index(drop=True)
+    data_frame = data_frame[cols]
+
+    present_time = arrow.now()
+    start_time = arrow.now().shift(days=-args.days_since_yday)
+    end_time = present_time.shift(days=-1)
+    file_name = "gh_data_{days}.csv".format(days='-'.join([start_time.format('YYYYMMDD'), end_time.format('YYYYMMDD')]))
+    data_frame.to_csv('/app/gh_data/{filename}'.format(filename=file_name), index=False)
```
Comment (arajkumar) [inline run_data_collector.py]: Actually you can directly dump the panda's DF into S3.

`data_frame.to_csv('s3://bucket')`

Even you can get rid of aws_utils.py :)

Reference: https://stackoverflow.com/a/56275519/1942688

File: scripts/entrypoint.sh (hunk 2 above)
Comment (arajkumar) [inline scripts/entrypoint.sh]: How to differentiate the ecosystem specific events if you combine all into one? This information is needed when generating a report.

**Review by arajkumar — CHANGES_REQUESTED** [review]

**Review by rootAvish — CHANGES_REQUESTED** [review]
Two comments here:   

1) At least the main flows of the source code need to be sanity tested.
2) If you do not want to test the AWS utils for example as that would require a lot of mocking etc., you're free to use the Rudra library for AWS and Bigquery utilities: https://github.com/fabric8-analytics/fabric8-analytics-rudra

File: scripts/entrypoint.sh [hunk 5]
```diff
@@ -3,5 +3,5 @@ set +ex
 
 ecosystem_list=("openshift" "knative" "kubevirt")
 
-# Run the data collector for given eco systems and no of days to colelct data
-python3 app/run_data_collector.py -eco-systems ${ecosystem_list[*]} -days ${DAYS};
+# Run the data collector for given eco systems and no of days to  collect data
+python3 run_data_collector.py -e ${ecosystem_list[*]} -d ${DAYS};
```
Comment (rootAvish) [inline scripts/entrypoint.sh:7]: It is better to specify absolute path of this script here.

File: requirements.in [hunk 6]
```diff
@@ -1,7 +1,8 @@
-boto3
 tqdm
 arrow
 pandas
 daiquiri
 google-cloud-bigquery
 lxml
+s3fs
+-e git+git://github.com/SohierDane/BigQuery_Helper#egg=bq_helper
```
Comment (rootAvish) [inline requirements.in]: @arajkumar I can see this last commit to this repository was in 2018, the big query API has changed quite a bit in that time. Are you sure you guys want to use it?

**Review by rootAvish — CHANGES_REQUESTED** [review]

File: src/utils/bq_utils.py (hunk 1 above)
Comment (rajusem) [reply src/utils/bq_utils.py]: now referring that package

File: scripts/entrypoint.sh (hunk 2 above)
Comment (rajusem) [reply scripts/entrypoint.sh]: corrected

File: run_data_collector.py (hunk 4 above)
Comment (rajusem) [reply run_data_collector.py]: now using s3 path directly with panda library

File: scripts/entrypoint.sh (hunk 2 above)
Comment (rajusem) [reply scripts/entrypoint.sh]: As discussed with you code is written such a way we cant differentiate for now, will be taking some story to work on it in future sprint.

File: scripts/entrypoint.sh (hunk 5 above)
Comment (rajusem) [reply scripts/entrypoint.sh:7]: we are copying run_data_collector.py to root path only and using the same.

File: requirements.in (hunk 6 above)
Comment (arajkumar) [reply requirements.in]: It is better to refer to the dependency instead of copying the code into our source tree. Also if the latest is incompatible with our implementation, use version pinning(of course here it can be achieved by sticking to particular commit).

File: tests/test_bq_data_collector.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
+import unittest
```
Comment (arajkumar) [inline tests/test_bq_data_collector.py]: Test file should follow same folder hierarchy of source tree

File: tests/test_helper.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import os
+
+import src.utils.bq_client_helper as bq_client_helper
+
+
+def get_file_absolute_path(relative_path):
+    working_dir = os.path.abspath(os.getcwd())
+    absolute_file_path = "{dir}{relative_path}".format(dir=working_dir, relative_path=relative_path)
```
Comment (arajkumar) [inline tests/test_helper.py]: Use os.path.join instead of string format

File: requirements.txt [hunk 9]
```diff
@@ -0,0 +1,42 @@
+#
+# This file is autogenerated by pip-compile
+# To update, run:
+#
+#    pip-compile requirements.in
+#
+-e git+git://github.com/SohierDane/BigQuery_Helper#egg=bq_helper  # via -r requirements.in
```
Comment (arajkumar) [inline requirements.txt]: Also instead of egg, you can install as a zip file(with commit/version locking),

https://github.com/SohierDane/BigQuery_Helper/archive/8615a7f6c1663e7f2d48aa2b32c2dbcb600a440f.zip

File: run_data_collector.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    :param data_frame: panda dataframe that has github data
+    :param days_since_yday: no of days we given in argument
+    """
+    present_time = arrow.now()
+    start_time = arrow.now().shift(days=-days_since_yday)
+    end_time = present_time.shift(days=-1)
+    file_name = "gh_data_{days}.csv".format(days='-'.join([start_time.format('YYYYMMDD'), end_time.format('YYYYMMDD')]))
+    _logger.info('Uploading Github data to S3 Bucket')
+    try:
+        data_frame.to_csv('s3://{bucket}/gh_data/{filename}'
```
Comment (arajkumar) [inline run_data_collector.py]: How to differentiate ecosystem specific events if all dependency events are dumped into single CSV.?

File: run_data_collector.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+def save_data_to_object_store(data_frame, days_since_yday):
+    """
+    Savethe github data to object s3 store
+    :param data_frame: panda dataframe that has github data
+    :param days_since_yday: no of days we given in argument
+    """
+    present_time = arrow.now()
+    start_time = arrow.now().shift(days=-days_since_yday)
+    end_time = present_time.shift(days=-1)
```
Comment (arajkumar) [inline run_data_collector.py]: Is it possible to expose `_get_query_date_range` and call it to get the date range instead of duplicating it?

File: tests/test_bq_data_collector.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+import src.utils.cloud_constants as cc
+import tests.test_helper as test_helper
+from src.bq_data_collector import BigQueryDataCollector
+
+
+class BigDataCollectorTestCase(unittest.TestCase):
+
+    @patch('src.utils.bq_client_helper.create_github_bq_client', return_value=MagicMock())
+    def test_get_gh_event_as_data_frame(self, _mock_bq_client):
```
Comment (arajkumar) [inline tests/test_bq_data_collector.py]: You may need to have a test for empty BQ response also.

**Review by arajkumar — CHANGES_REQUESTED** [review]
I don't see pytest as a dependency in requirements.txt

File: tests/test_bq_data_collector.py (hunk 7 above)
Comment (rajusem) [reply tests/test_bq_data_collector.py]: done

File: tests/test_helper.py (hunk 8 above)
Comment (rajusem) [reply tests/test_helper.py]: done

File: requirements.txt (hunk 9 above)
Comment (rajusem) [reply requirements.txt]: I tried to use zip but somehow not able use it after installing it. Tried to use new venv but not helping. So not using zip but added commit version in to git url with egg

File: run_data_collector.py (hunk 10 above)
Comment (rajusem) [reply run_data_collector.py]: added logic for the same now. Based on rep_name it will update ecosystem.

File: run_data_collector.py (hunk 11 above)
Comment (rajusem) [reply run_data_collector.py]: Now using same function that returns tuple, using those values as required.

File: tests/test_bq_data_collector.py (hunk 12 above)
Comment (rajusem) [reply tests/test_bq_data_collector.py]: aded empty BQ response test case

**Comment by rajusem** [comment]
> I don't see pytest as a dependency in requirements.txt
 
@arajkumar We are not using pytest, using unittest module which come with python

File: src/bq_data_collector.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+        """
+        if repo_name in self._golang_repo:
+            return 'golang'
+        elif repo_name in self._knative_repo:
+            return 'knative'
+        elif repo_name in self._kubevirt_repo:
+            return 'kubevirt'
+        else:
+            return None
+
```
Comment (arajkumar) [inline src/bq_data_collector.py]: Few dependencies(e.g. golang/go) are common in all 3 ecosystems, in that case your latest implementation will end up tagging it wrong. I guess you may need to change the implementation or fallback to previous one.

Comment (rajusem) [reply src/bq_data_collector.py]: @arajkumar : in this case we can add ecosystem as  comm separated values Ex "golang,knative" etc. What do you think?

Comment (arajkumar) [reply src/bq_data_collector.py]: This might work when the no.of ecosystems are low, but when it explodes it will be come cumbersome., but as of now you can go ahead with CSV for ecosystem field.

Comment (rajusem) [reply src/bq_data_collector.py]: As discussed added support for multiple eco-system. It will stored into 'ecosystem' column with comma separated.  Ex : "golang,knative"

File: src/bq_data_collector.py [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+_logger = daiquiri.getLogger(__name__)
+
+
+class BigQueryDataCollector:
+    def __init__(self, repos: List[str], bq_credentials_path: str = '', days: int = 3):
+        self._bq_client = BigQueryDataCollector._get_bq_client(bq_credentials_path)
+        self._golang_repo = bq_client_helper.get_gokube_trackable_repos(repo_dir=cc.GOKUBE_REPO_LIST)
+        self._knative_repo = bq_client_helper.get_gokube_trackable_repos(repo_dir=cc.KNATIVE_REPO_LIST)
+        self._kubevirt_repo = bq_client_helper.get_gokube_trackable_repos(repo_dir=cc.KUBEVIRT_REPO_LIST)
+        self._init_query_param(repos, days)
```
Comment (arajkumar) [inline src/bq_data_collector.py]: if you go ahead with CSV, this won't scale better if we wanna track more other ecosystems. It is better to use dictionary/array based approach to keep details instead of dedicated instance variable for each ecosystem.

Comment (rajusem) [reply src/bq_data_collector.py]: Added dictionary with all eco-system and repo names.

File: src/utils/data_assets/repo-list.json [hunk 15]
```diff
@@ -0,0 +1,1089 @@
+[
+  {
+    "eco-system": "openshift",
```
Comment (arajkumar) [inline src/utils/data_assets/repo-list.json]: Rename it to ecosystem

File: src/utils/bq_client_helper.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+        query = query.replace(i, j)
+    return query
+
+
+def get_eco_system_with_repo_list():
+    """
+    Read the repo-list.json file and make a dictionary that contains list with ecosystem name and repos
+    """
+    with open(cc.REPO_LIST) as file:
+        json_array = json.load(file)
```
Comment (arajkumar) [inline src/utils/bq_client_helper.py]: Variable name is misleading. `json.load` returns python dictionary not array. Rename it to ecosystems or something similar

File: src/utils/bq_client_helper.py [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+        eco_system_with_repo = []
+        for item in json_array:
```
Comment (arajkumar) [inline src/utils/bq_client_helper.py]: It would be nice if ecosystem specific transformations are abstracted into a class.
```
class Ecosystem:
    def __init__(self, ecosystem_dict):
        self.name = ...
        self.repo_names = ...
```

File: src/bq_data_collector.py [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+
+warnings.simplefilter(action='ignore', category=FutureWarning)
+warnings.simplefilter(action='ignore', category=Warning)
+
+daiquiri.setup(level=logging.INFO)
+_logger = daiquiri.getLogger(__name__)
+
+
+class BigQueryDataCollector:
+    def __init__(self, repos: List[str], bq_credentials_path: str = '', days: int = 3):
```
Comment (arajkumar) [inline src/bq_data_collector.py]: It takes list of ecosystems, not repos

File: src/utils/bq_client_helper.py [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+    for i, j in params_dict.items():
+        query = query.replace(i, j)
+    return query
+
+
+def get_eco_system_with_repo_list():
+    """
+    Read the repo-list.json file and make a dictionary that contains list with ecosystem name and repos
+    """
+    with open(cc.REPO_LIST) as file:
```
Comment (arajkumar) [inline src/utils/bq_client_helper.py]: Do dependency injection similar like `bq_credentials_path` , it will help you to do better unit testing without mocking `cc.REPO_LIST`.

File: src/utils/bq_client_helper.py (hunk 17 above)
Comment (arajkumar) [inline src/utils/bq_client_helper.py]: you can have multiple return value of dict based iterator, so
```
for ecosytem, urls in json_array:
...
```

File: src/utils/bq_client_helper.py [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+        json_array = json.load(file)
+        eco_system_with_repo = []
+        for item in json_array:
+            repo_names = get_repos_names(item['urls'])
+            eco_system_with_repo.append({'eco-system': item['eco-system'], 'repo-names': repo_names})
```
Comment (arajkumar) [inline src/utils/bq_client_helper.py]: keep consistent naming for key and variable. (i.e. change repo-name to repo_names)

File: src/bq_data_collector.py [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+        else:
+            msg = 'Given eco-system "{eco}" is not supported'.format(eco=eco_system)
+            _logger.error(msg)
+            return []
+
+    def _get_repo_list(self, eco_systems: List[str]) -> List[str]:
+        """
+        Get Repo list based on eco system passed
+        """
+        repo_names = list()
```
Comment (arajkumar) [inline src/bq_data_collector.py]: repo_names type should be set to avoid duplicate, not list. 

e.g. `ecosystem:a, urls:[xyz], ecosystem:b, urls:[xyz]` - In this case query should have only one entry that is xyz.

File: src/bq_data_collector.py [hunk 22]
```diff
@@ -0,0 +1,246 @@
...
+warnings.simplefilter(action='ignore', category=Warning)
+
+daiquiri.setup(level=logging.INFO)
+_logger = daiquiri.getLogger(__name__)
+
+
+class BigQueryDataCollector:
+    def __init__(self, repos: List[str], bq_credentials_path: str = '', days: int = 3):
+        self._bq_client = BigQueryDataCollector._get_bq_client(bq_credentials_path)
+        self._repo_list = bq_client_helper.get_eco_system_with_repo_list()
```
Comment (arajkumar) [inline src/bq_data_collector.py]: _repo_list should be an `map` with ecosystem as key for easy retrieval. That is how data is modelled right?

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: src/utils/data_assets/repo-list.json (hunk 15 above)
Comment (rajusem) [reply src/utils/data_assets/repo-list.json]: done

File: src/utils/bq_client_helper.py (hunk 16 above)
Comment (rajusem) [reply src/utils/bq_client_helper.py]: As we are storing json with list of object,  json.load returns list only not dictionary. We are using this list to create dictionary for later use.

File: src/utils/bq_client_helper.py (hunk 17 above)
Comment (rajusem) [reply src/utils/bq_client_helper.py]: As discussed not adding this class now as we are using dictionary other places.

File: src/bq_data_collector.py (hunk 18 above)
Comment (rajusem) [reply src/bq_data_collector.py]: corrected

File: src/utils/bq_client_helper.py (hunk 19 above)
Comment (rajusem) [reply src/utils/bq_client_helper.py]: now passing url into method and reading it.

File: src/utils/bq_client_helper.py (hunk 17 above)
Comment (rajusem) [reply src/utils/bq_client_helper.py]: As its list (not dictionary) we cant do like this.

File: src/utils/bq_client_helper.py (hunk 20 above)
Comment (rajusem) [reply src/utils/bq_client_helper.py]: done

File: src/bq_data_collector.py (hunk 21 above)
Comment (rajusem) [reply src/bq_data_collector.py]: changed it to set

File: src/bq_data_collector.py (hunk 22 above)
Comment (rajusem) [reply src/bq_data_collector.py]: as discussed now _repo_list will come as dictionary.

File: requirements.in [hunk 23]
```diff
@@ -0,0 +1,8 @@
+tqdm
```
Comment (rootAvish) [inline requirements.in]: I think this dependency is not required.

**Review by rootAvish — COMMENTED** [review]
LGTM, one very minor change.

File: requirements.in (hunk 23 above)
Comment (rajusem) [reply requirements.in]: Yes its not require, removed the same.

File: src/bq_data_collector.py [hunk 24]
```diff
@@ -19,31 +18,32 @@
 
 
 class BigQueryDataCollector:
-    def __init__(self, repos: List[str], bq_credentials_path: str = '', days: int = 3):
+    def __init__(self, ecosystems: List[str], bq_credentials_path: str = '', days: int = 3):
         self._bq_client = BigQueryDataCollector._get_bq_client(bq_credentials_path)
-        self._repo_list = bq_client_helper.get_eco_system_with_repo_list()
-        self._init_query_param(repos, days)
+        self._repo_list = bq_client_helper.get_eco_system_with_repo_list(cc.REPO_LIST)
```
Comment (arajkumar) [inline src/bq_data_collector.py]: Repo list must be argumented to BigQueryDataCollector constructor similar to bq_credentials_path.

**Review by arajkumar — CHANGES_REQUESTED** [review]

File: src/bq_data_collector.py (hunk 24 above)
Comment (rajusem) [reply src/bq_data_collector.py]: Now taking it as param

File: tests/src/test_bq_data_collector.py [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+    def test_get_repo_by_eco_system_invalid_name(self):
+        repo = self._bq_data_collector._get_repo_by_eco_system('invalid-repo')
+        self.assertEqual(0, len(repo))
+
+    def test_init_query_param(self):
+        present_time = arrow.now()
+        start_time = present_time.shift(days=-self._no_of_days)
+        end_time = present_time.shift(days=-1)
+        last_n_days = [dt.format('YYYYMMDD') for dt in arrow.Arrow.range('day', start_time, end_time)]
+        day_list = '(' + ', '.join(["'" + d + "'" for d in [item[2:] for item in last_n_days]]) + ')'
```
Comment (rootAvish) [inline tests/src/test_bq_data_collector.py]: Can you join this with `.format()`, will improve readability?

File: src/bq_data_collector.py [hunk 26]
```diff
@@ -0,0 +1,242 @@
...
+        if data_frame.empty:
+            _logger.warn('Nothing to save')
+        else:
+
+            last_n_days, start_time, end_time = self._get_query_date_range(days_since_yday)
+            file_name = "gh_data_{days}.csv".format(
+                days='-'.join([start_time.format('YYYYMMDD'), end_time.format('YYYYMMDD')]))
+            _logger.info('Uploading Github data to S3 Bucket')
+            try:
+                data_frame.to_csv(
```
Comment (rootAvish) [inline src/bq_data_collector.py:235]: Does this directly upload to S3?

File: tests/src/test_helper.py [hunk 27]
```diff
@@ -0,0 +1,31 @@
...
+
+
+def get_sample_query_param():
+    """
+    Build sample query param used in BigQueryDataCollector class
+    """
+    repo_names = get_sample_repo_names()
+    day_list = ['200303', '200304']
+    year_prefix = '20*'
+    query_params = {'{year_prefix_wildcard}': year_prefix,
```
Comment (rootAvish) [inline tests/src/test_helper.py:29]: Here as well, can you join the strings with `.format()`?

**Review by rootAvish — COMMENTED** [review]
Just one-two comments otherwise looks good.

File: tests/src/test_bq_data_collector.py (hunk 25 above)
Comment (rajusem) [reply tests/src/test_bq_data_collector.py]: done

File: src/bq_data_collector.py (hunk 26 above)
Comment (rajusem) [reply src/bq_data_collector.py:235]: yes its uploading to s3

File: tests/src/test_helper.py (hunk 27 above)
Comment (rajusem) [reply tests/src/test_helper.py:29]: done

**Review by rootAvish — APPROVED** [review]

**Review by arajkumar — APPROVED** [review]
lgtm
